addons/sale_stock/models/sale_order.py

```diff
@@ -98,6 +99,13 @@ def _select_expected_date(self, expected_dates):
             return super()._select_expected_date(expected_dates)
         return max(expected_dates)
 
+    @api.constrains('warehouse_id', 'state', 'order_line')
+    def _check_warehouse(self):
+        """ Ensure that the warehouse is set in case of storable products """
+        for order in self:
+            if order.state not in ['draft', 'cancel'] and any(product.type in ['product', 'consu'] for product in order.order_line.product_id) and not order.warehouse_id:
```

Comment:
I'm not sure it's the best way to do it. You could rather select all order without wh and then do a gobal search on line in them with goods type (product, consu doesn't exists anymore)

Reply:
and in general try to order the condition by time consuming operation. It's not needed to brwose all the list if order.warehouse_id is set